Add support for templates in config

justenwalker opened the pull request:
This is my first pass at supporting templates in configurations:

- Config is parsed as a Go template before being unmarshaled from JSON.
- Environment variables are available for interpolation. ie: `{{ .ENV_VAR_NAME }}`

For #16

tgross commented:
There's some behavior in failure modes that seems to undo some of the UX improvements that we had in https://github.com/joyent/containerbuddy/pull/45. If I change the example application so that the Nginx port looks like this:

```
{
  "consul": "consul:8500",
  "services": [
    {
      "name": "nginx",
      "port": {{ .NGINX_PORT }},
      "interfaces": ["eth0"],
      "health": "/usr/bin/curl --fail -s http://localhost/health.txt",
      "poll": 10,
      "ttl": 25
    }
...
```

And then don't add `NGINX_PORT=80` to my `docker-compose-local.yml` I get this somewhat unintuitive-looking error:

> 2015/12/14 18:33:48 Could not parse configuration: invalid character ',' looking for beginning of value

It works just fine if I do pass the variable. But the failure mode is cryptic.

tgross commented on README.md:
```diff
@@ -126,6 +126,22 @@ All executable fields, such as `onStart` and `onChange`, accept both a string or
 
+### Template Configuration
+
+Containerbuddy configuration has template support, such as substituting environment variables in the config file. This makes it possible to use the same configuration and alter the behavior per environment. Full documentation about the template language used can be found in the [Go text/template Docs](https://golang.org/pkg/text/template/).
```
I feel like we should be a bit more straightforward about how environment variables are interpreted. The `text/template` docs are accurate but are dense and talk about "fields" (which is an implementation detail) but what we're really trying to say is "if you set an environment variable `FOO=BAR` then you can read it into the template like `{{ .FOO }}`". We should make that common case front-and-center of this section and then follow up with something along the lines of "oh and this is Go `text/template` language if you want more".

justenwalker commented:
To your comment about the cryptic error message.  I'm not sure how to resolve this, since it seems to happen inside the `json.Unmarshal` - where the checks happen after those values have already been parsed.

Any ideas on where I should look to resolve this?

tgross commented:
> I'm not sure how to resolve this, since it seems to happen inside the json.Unmarshal
> The UX checks happen after those values have already been parsed.

Hm. Maybe we should be using `missingkey=error` in that case? That way we get an error at the point it happens and can give the user a sensible error message about there being a missing value?

justenwalker commented:
I'm on board with the idea of making it an explicit error. 

I liked the blank value first because it closely mirrored what I expect to happen in bash scripts. If we leave it as `missingkey=zero` it will allow people who are using containerbuddy to have optional environment variables in their config more easily.

misterbisson commented:
> I liked the blank value first because it closely mirrored what I expect to happen in bash scripts. If we leave it as missingkey=zero it will allow people who are using containerbuddy to have optional environment variables in their config more easily.

I think we can demo some `Dockerfile` patterns to go with this that include default values for environment vars, see [example](https://github.com/joyent/mesos-dockerfiles/blob/master/agent/Dockerfile#L28-L31).

tgross commented:
> If we leave it as missingkey=zero it will allow people who are using containerbuddy to have optional environment variables in their config more easily.

I think unfortunately that's only going to work for string variables, because they get quoted and so the JSON parser has at the very least `""` to work with.

justenwalker commented:
In terms of environment variables - they are always strings, no?

tgross commented:
> In terms of environment variables - they are always strings, no?

Right, but they aren't strings in the JSON, which appears to be the problem here. The zero-value for the environment variable is indeed `""`, but that results in the JSON:

```
{
  "consul": "consul:8500",
  "services": [
    {
      "name": "nginx",
      "port": ,
      "interfaces": ["eth0"],
      "health": "/usr/bin/curl --fail -s http://localhost/health.txt",
      "poll": 10,
      "ttl": 25
    }
...
```

tgross commented:
Sorry... missed this comment prior to my one above somehow.

> Perhaps we can add a default function that can be used to provide sane defaults to numeric values?

Strictly speaking we can do this already with `{{if .NGINX_PORT }}{{ .NGINX_PORT }}{{else}}80{{end}}` but I in keeping with the "this should be easy to understand" theme, I like the idea of having a pipeline function that does it better.

justenwalker commented:
@tgross Still wont solve the issue entirely though, since if the environment variable is omitted - it will fail with this cryptic error. Maybe we can parse out the syntax error message and make some more meaningful error in this case (see https://groups.google.com/forum/#!topic/golang-nuts/fizimmXtVfc)

misterbisson commented:
Would it be possible to pre-parse the config file to find the referenced env vars and test their availability?

tgross commented:
@misterbisson that's what I proposed [here](https://github.com/joyent/containerbuddy/pull/46#issuecomment-164531709)

tgross commented:
> Still won't solve the issue entirely though, since if the environment variable is omitted - it will fail with this cryptic error

Sure, that's true. We're stuck with the possibility of misconfiguration regardless of what we do, and with @misterbisson's suggestion about Dockerfile defaults and the existing template language options there are options for the user to reduce errors as it is. So doing our darndest to return a sensible error message seems like the priority.

misterbisson commented:
> that's essentially what I proposed

Sorry, I didn't understand [your comment](https://github.com/joyent/containerbuddy/pull/46#issuecomment-164531709) in that context, thanks for clarifying @tgross.

justenwalker commented:
I pushed up a change that should:

1. Allow default values with pipeline {{ .ENV_VAR | default "hello!" }}
2. Give us better error messages on json parse errors:

```
2015/12/14 21:48:44 Parse error on line 6: invalid character ',' looking for beginning of value
      "port": ,
              ^
```

Line number code based off that google groups discussion that I mentioned above.

justenwalker commented:
Updated the implementation of the JSON error pretty-printer

Implementation is a little easier to follow and maintain with Scanners. 
Plus I added some additional context to the error message:


**Examples**

```
Parse error at line:col [1:1]: invalid character '<' looking for beginning of value
    1: <>
-------^

Parse error at line:col [2:13]: invalid character ',' looking for beginning of value
    1: {
    2:     "test": ,
-------------------^

Parse error at line:col [5:5]: invalid character 't' looking for beginning of object key string
    4:     "test3": false,
    5:     test2: "hello"
-----------^
```

tgross commented on out.txt:
```diff
@@ -0,0 +1,75 @@
+docker rm -f containerbuddy_consul || true
```
was this file committed accidentally?

justenwalker replied:
yes - removed

tgross commented on src/containerbuddy/config.go:
```diff
@@ -271,25 +273,69 @@ func parseConfig(configFlag string) (*Config, error) {
+	return line, int(col), fmt.Sprintf("%s%s%s", prevLine, thisLine, highlight)
+}
+
+func newJSONParseError(js []byte, syntax *json.SyntaxError) error {
```
This is just me being picky, but can we move this function above `highlightError`? The only call chain for this is `unmarshalConfig --> newJSONParseError --> highlightError` so legibility is helped if you don't have to jump down and then back up again.

justenwalker replied:
Done

tgross commented:
Nice work. LGTM.
